pkg/apis/extensions/v1alpha1/types.go

```diff
@@ -0,0 +1,198 @@
+	ExtensionRecord *corev1.ObjectReference `json:"extensionRecord,omitempty"`
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Extension is the CRD type for a request to add an extension to Crossplane.
+// +k8s:openapi-gen=true
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="STATUS",type="string",JSONPath=".status.conditions[0].type"
+// +kubebuilder:printcolumn:name="DESCRIPTION",type="string",JSONPath=".spec.description"
+// +kubebuilder:printcolumn:name="VERSION",type="string",JSONPath=".spec.version"
+// +kubebuilder:printcolumn:name="AGE",type="date",JSONPath=".metadata.creationTimestamp"
+type Extension struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ExtensionSpec   `json:"spec,omitempty"`
+	Status ExtensionStatus `json:"status,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ExtensionList contains a list of Extension
+type ExtensionList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []Extension `json:"items"`
+}
+
+// ExtensionSpec specifies details about an extension that has been added to Crossplane
+type ExtensionSpec struct {
+	AppMetadataSpec `json:",inline"`
+	CRDs            CRDList         `json:"customresourcedefinitions,omitempty"`
+	Controller      ControllerSpec  `json:"controller,omitempty"`
+	Permissions     PermissionsSpec `json:"permissions,omitempty"`
+}
+
+// ExtensionStatus defines the observed state of Extension
+type ExtensionStatus struct {
+	corev1alpha1.ConditionedStatus
+}
+
+// AppMetadataSpec defines metadata about the extension application
+type AppMetadataSpec struct {
+	Title       string            `json:"title,omitempty"`
+	Description string            `json:"description,omitempty"`
+	Version     string            `json:"version,omitempty"`
+	Icons       []IconSpec        `json:"icons,omitempty"`
+	Maintainers []ContributorSpec `json:"maintainers,omitempty"`
+	Owners      []ContributorSpec `json:"owners,omitempty"`
+	Company     string            `json:"company,omitempty"`
+	Keywords    []string          `json:"keywords,omitempty"`
+	Links       []LinkSpec        `json:"links,omitempty"`
+	License     string            `json:"license,omitempty"`
+}
+
+// CRDList is the full list of CRDs that this extension owns and depends on
+type CRDList struct {
+	// Owned is the list of CRDs that this extension defines and owns
+	Owned []CRDSpec `json:"owns,omitempty"`
+
+	// DependsOn is the list of CRDs that this extension depends on. This data drives the
+	// dependency resolution process.
+	DependsOn []CRDSpec `json:"dependsOn,omitempty"`
+}
+
+// NewCRDList creates a new CRDList with its members initialized.
+func NewCRDList() *CRDList {
+	return &CRDList{
+		Owned:     []CRDSpec{},
+		DependsOn: []CRDSpec{},
+	}
+}
+
+// CRDSpec defines a specific CRD
+type CRDSpec struct {
```

Comment:
Any reason why we wouldn't want to use: `k8s.io/apimachinery/pkg/apis/meta/v1/types.go`
```go
type TypeMeta struct {
	// Kind is a string value representing the REST resource this object represents.
	// Servers may infer this from the endpoint the client submits requests to.
	// Cannot be updated.
	// In CamelCase.
	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#types-kinds
	// +optional
	Kind string `json:"kind,omitempty" protobuf:"bytes,1,opt,name=kind"`

	// APIVersion defines the versioned schema of this representation of an object.
	// Servers should convert recognized schemas to the latest internal value, and
	// may reject unrecognized values.
	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#resources
	// +optional
	APIVersion string `json:"apiVersion,omitempty" protobuf:"bytes,2,opt,name=apiVersion"`
}
```
instead of this `CRDSpec`?

Reply:
I think we could actually standardize on the `TypeMeta` upstream type, that captures everything we have individually defined here and there are helper methods upstream to parse into individual fields if we need.

So instead of our our `CRDSpec`:
```yaml
customresourcedefinitions:
   owns:
   - name: rediscluster.crossplane.redislabs.com
     kind: RedisCluster
     group: crossplane.redislabs.com
     version: v1alpha1
```

We'd have:
```yaml
customresourcedefinitions:
   owns:
   - kind: RedisCluster
     apiVersion: crossplane.redislabs.com/v1alpha1
```

Which is more inline with the information usually provided when creating an instance of a CRD.